tests/e2e/tools.go

```diff
@@ -0,0 +1,207 @@
+/*
+* Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+ */
+
+package e2e
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"text/template"
+	"time"
+
+	"golang.org/x/crypto/ssh"
+)
+
+var dockerInstallTemplate = `
+#! /usr/bin/env bash
+set -xe
+
+: ${IMAGE:={{.Image}}}
+
+# Create a temporary directory
+TEMP_DIR="/tmp/ctk_e2e.$(date +%s)_$RANDOM"
+mkdir -p "$TEMP_DIR"
+
+# Given that docker has an init function that checks for the existence of the
+# nvidia-container-toolkit, we need to create a symlink to the nvidia-container-runtime-hook
+# in the /usr/bin directory.
+# See https://github.com/moby/moby/blob/20a05dabf44934447d1a66cdd616cc803b81d4e2/daemon/nvidia_linux.go#L32-L46
+sudo ln -s "$TEMP_DIR/toolkit/nvidia-container-runtime-hook" /usr/bin/nvidia-container-runtime-hook
+
+docker run --pid=host --rm -i --privileged	\
+	-v /:/host	\
+	-v /var/run/docker.sock:/var/run/docker.sock	\
+	-v "$TEMP_DIR:$TEMP_DIR"	\
+	-v /etc/docker:/config-root	\
+	${IMAGE}	\
+	--root "$TEMP_DIR"	\
+	--runtime=docker	\
+	--config=/config-root/daemon.json	\
+	--driver-root=/	\
+	--no-daemon	\
+	--restart-mode=systemd
+`
+
+type Installer struct {
+	Image    string
+	Template string
+
+	SshKey     string
+	SshUser    string
+	RemoteHost string
+}
+
+func newInstaller(templateStr, image, sshKey, sshUser, remoteHost string) (Installer, error) {
+	i := Installer{
+		Image:      image,
+		Template:   templateStr,
+		SshKey:     sshKey,
+		SshUser:    sshUser,
+		RemoteHost: remoteHost,
+	}
+
+	return i, nil
+}
+
+func (i *Installer) Install() error {
+	// Parse the combined template
+	tmpl, err := template.New("dockerScript").Parse(i.Template)
+	if err != nil {
+		return fmt.Errorf("error parsing template: %w", err)
+	}
+
+	// Execute the template
+	var renderedScript bytes.Buffer
+	err = tmpl.Execute(&renderedScript, i)
+	if err != nil {
+		return fmt.Errorf("error executing template: %w", err)
+	}
+
+	var s Script
+	switch {
+	case i.RemoteHost == "":
+		s = localRun{}
+	default:
+		s = remoteRun{i.SshKey, i.SshUser, i.RemoteHost}
+	}
+
+	_, err = s.Run(renderedScript.String())
+	return err
+}
+
+type Script interface {
+	Run(script string) (string, error)
+}
+
+type localRun struct{}
+type remoteRun struct {
```

Comment:
```suggestion

// A remoteRunner runs scripts over SSH.
type remoteRunner struct {
```
(also wondering whether `sshRunner` isn't more accurate).